### PR TITLE
rUDP Update

### PR DESCRIPTION
**Pull request by hoffmabc**
Ok have at it @Renelvon @yagoulas @gubatron 

One other thing to note is that I have hardcoded in the only compatible seed node which is seed2.openbazaar.org. Once the other seeds update then we can add them to the list in a later commit.

### CODE REVIEW

**Comment by hoffmabc** [comment]
This was by far a horrible rebase because there were some definite issues with the DHT refactor because the method and variable names were not thoroughly updated across the codebase.

**Comment by gubatron** [comment]
Trying to pull latest changes, but I'm having issues:

```
Angels-MBP:OpenBazaar gubatron$ git reset --hard HEAD
HEAD is now at a6393eb Lint fixes
Angels-MBP:OpenBazaar gubatron$ git pull hoffmabc brian-rudp 
From https://github.com/hoffmabc/OpenBazaar
 * branch            brian-rudp -> FETCH_HEAD
Auto-merging rudp/sender.py
CONFLICT (add/add): Merge conflict in rudp/sender.py
Auto-merging rudp/receiver.py
CONFLICT (add/add): Merge conflict in rudp/receiver.py
Auto-merging rudp/pendingpacket.py
CONFLICT (add/add): Merge conflict in rudp/pendingpacket.py
Auto-merging rudp/packetsender.py
CONFLICT (add/add): Merge conflict in rudp/packetsender.py
Auto-merging rudp/packet.py
CONFLICT (add/add): Merge conflict in rudp/packet.py
Auto-merging rudp/mediator.py
CONFLICT (add/add): Merge conflict in rudp/mediator.py
Auto-merging rudp/linkedlist.py
CONFLICT (add/add): Merge conflict in rudp/linkedlist.py
Auto-merging rudp/helpers.py
CONFLICT (add/add): Merge conflict in rudp/helpers.py
Auto-merging rudp/connection.py
CONFLICT (add/add): Merge conflict in rudp/connection.py
Auto-merging requirements.txt
Auto-merging node/ws.py
CONFLICT (content): Merge conflict in node/ws.py
Auto-merging node/trust.py
CONFLICT (content): Merge conflict in node/trust.py
Auto-merging node/transport.py
CONFLICT (content): Merge conflict in node/transport.py
Auto-merging node/orders.py
CONFLICT (content): Merge conflict in node/orders.py
Auto-merging node/openbazaar_daemon.py
Auto-merging node/market.py
CONFLICT (content): Merge conflict in node/market.py
Auto-merging node/dht.py
CONFLICT (content): Merge conflict in node/dht.py
Auto-merging html/partials/settings.html
Auto-merging html/css/dark.css
Auto-merging configure.sh
Auto-merging Vagrantfile
CONFLICT (content): Merge conflict in Vagrantfile
Automatic merge failed; fix conflicts and then commit the result.
```

**Comment by gubatron** [comment]
nvm, deleted the local branch and fetched it again.

**Comment by hoffmabc** [comment]
Ok @Renelvon @gubatron @yagoulas I think the Travis build should come up clean.

**Comment by Renelvon** [comment]
:+1: Will review tomorrow.

**Comment by hoffmabc** [comment]
Fuck yeah.

**Comment by Renelvon** [comment]
Places still containing references to ZeroMQ:
* `ansible-provisioning/openbazaar_linux.yml`
* `installers/windows/setup.py`
* `installers/ubuntu/DEBIAN/control`
* `features/environment.py`
* `features/steps/CryptoTransportLayer.py`
* `node/transport.py`
* `util/docker/logelk/ob/Dockerfile`
* `configure.sh`

**Comment by hoffmabc** [comment]
OK @Renelvon removed all the references.

**Comment by hoffmabc** [comment]
@renelvon any thoughts on this Travis error? It looks like the Obelisk library we refer to imports the Bitcoin library, which in turn uses python-ecdsa and calls the Curve class with the wrong constructor arguments. But this new argument was added in like July.

**Comment by yagoulas** [comment]
@hoffmabc: in the coveralls report rudp isn't included in the code that is taken in to account. Could you fix this as well?

**Comment by hoffmabc** [comment]
Yeah didn't think about that. I will update. When are we going to get a good enough review here?

**Comment by hoffmabc** [comment]
@yagoulas I just checked and the only thing that was missing the rudp folder was the bandit file so I updated. Maybe I just don't know where coveralls reads?

File: html/controllers/market.js [hunk 1]
```diff
@@ -247,7 +248,7 @@ angular.module('app')
                 /* get index if peer is already known */
                 var index = [-1].concat($scope.peers).reduce(
                     function(previousValue, currentValue, index, array) {
-                        return currentValue.uri == msg.uri ? index : previousValue;
+                        return currentValue.guid == msg.guid ? index : previousValue;
```
Comment (gubatron) [inline html/controllers/market.js]: Maybe these also need to be fixed:

```
html/controllers/market.js:52                        'uri': msg.uri,
html/controllers/market.js:288                    return element.uri != msg.uri;
util/docker/logelk/elk/kibana-ob.json:100         "field": "obmsg.uri",
```

File: node/connection.py [hunk 2]
```diff
@@ -1,136 +1,161 @@
...
-                self.socket.connect(self.address)
-            except zmq.ZMQError as err:
-                self.log.error('Bad URI %s', err)
+        self.ee = EventEmitter()
+        self.sock = peer_socket
+
+        self.init_packetsender()
+
+    def init_packetsender(self):
+        self._packet_sender = PacketSender(
```
Comment (gubatron) [inline node/connection.py]: it'd probably be good to re-arrange the order of the parameters of this and other constructor to be consistent with one another and avoid bugs due to programmer assumptions (not everyone uses IDEs with code-completion features):

```
PeerConnection(transport, hostname, port, nickname, peer_socket)
CryptoPeerConnection(transport, hostname, port, pub, guid, nickname, sin=None, rudp_connection, peer_socket)

PacketSender(sock, hostname, port, transport)

PeerListener(hostname, port, guid, data_cb)
CryptoPeerListener(hostname, port, pubkey, secret, guid, data_cb)
```

I'd suggest the following refactor which makes the tuple (hostname, port) into a `InetSocketAddress`:
```
class InetSocketAddress:
   def __init__(self, hostname, port):
       self.hostname = hostname
       self.port = port
```

then, we substitute and re-arrange:
```python
PeerConnection(transport, socket_address, nickname, peer_socket)
CryptoPeerConnection(transport, socket_address, pub, guid, nickname, sin=None, rudp_connection, peer_socket)
PacketSender(transport, socket_address, sock) #this one needs major re-arrangement

PeerListener(socket_address, guid, data_cb)
CryptoPeerListener(socket_address, pubkey, secret, guid, data_cb)
```

File: node/connection.py [hunk 3]
```diff
@@ -1,136 +1,161 @@
...
 
-        self.stream.on_recv_stream(cb)
 
 
 class CryptoPeerConnection(GUIDMixin, PeerConnection):
 
-    def __init__(self, transport, address, pub=None, guid=None, nickname="",
-                 sin=None):
+    def __init__(self, transport, hostname, port, pub=None, guid=None, nickname="",
+                 sin=None, rudp_connection=None, peer_socket=None):
```
Comment (gubatron) [inline node/connection.py]: the `rudp_connection` parameter is unused, remove.

File: node/connection.py [hunk 4]
```diff
@@ -1,136 +1,161 @@
...
-        # Convert URI over
-        url = urlparse(address)
-        self.ip = url.hostname
-        self.port = url.port
+        self.setup_emitters()
 
-        self.sin = sin
-        self.address = "tcp://%s:%s" % (self.ip, self.port)
+        @self._rudp_connection._sender.ee.on('timeout')
+        def on_timeout(data):  # pylint: disable=unused-variable
```
Comment (gubatron) [inline node/connection.py]: this function is unused, remove.

Comment (gubatron) [inline node/connection.py]: actually, nv,, I believe it's used as an emitter but my IDE doesn't understand the decorator.

Comment (gubatron) [inline node/connection.py]: I suggest this is moved inside `setup_emitters`, which is called one line above.

File: node/connection.py [hunk 5]
```diff
@@ -1,136 +1,161 @@
...
-        )
+            try:
+                # payload = base64.b64decode(msg.get('payload'))
+                payload = msg.get('payload').decode('hex')
+                self.transport.listener._on_raw_message(payload)
+            except Exception as e:
+                self.log.debug('not yet %s', e)
+                self.transport.listener._on_raw_message(msg.get('payload'))
+
+    def start_handshake(self, initial_handshake_cb=None):
```
Comment (gubatron) [inline node/connection.py]: remove

**Comment by gubatron** [comment]
On `node/connection.py` `PeerListener::set_ip_address` there's a chance a call to `self.listen()` might never be executed as `self.stream` might be null:

```
    def set_ip_address(self, new_ip):
        self.hostname = new_ip
        if not self.is_listening:
            return

        try:
            self.stream.close()  #this right here
            self.listen()
        except Exception as e:
            self.log.error('[Requests] error: %s', e)
```

I'd rewrite to probably return a boolean and let the api user know wether this succeeded or not to:
```
    def set_ip_address(self, new_ip):
        self.hostname = new_ip
        if not self.is_listening:
            return False

        if self.stream is not None:
            self.stream.close()

        try:
            self.listen()
        except Exception as e:
            self.log.error('[Requests] error: %s', e)
            return False
        return True
```

```

File: node/connection.py [hunk 6]
```diff
@@ -302,9 +329,13 @@ def is_handshake(message):
         :param message: serialized JSON
         :return: True if proper handshake message
         """
+        if type(message) is 'dict':
```
Comment (gubatron) [inline node/connection.py]: this logic construct is quite weird, is the dictionary supposed to have a `True` inside `message['type']`?

Other wise, a json object is created, and we just ask if the message has a `type` field in it.

This seems very flimsy logic to determine wether or not this is a handshake.

First, there's the issue that it takes 2 different forms of input on the same method, a dictionary or a JSON string.

And then, the validation seems to be very weak, therefore making this a very likely source of bugs, errors or security vulnerabilities.

Comment (gubatron) [inline node/connection.py]: I suggest we have a `type=handshake` of sorts.

And taking out the JSON deserialization to a single place where it's probably supposed to happen, from which then on we work with message objects.

Comment (gubatron) [inline node/connection.py]: that place looks like `_on_raw_message`, perhaps everything that comes here should be first JSON.deserialized, then once we have a dict, we can do the checking in a consistent manner.

**Comment by gubatron** [comment]
I suggest that `CryptoPeerConnection` initializes in its constructor a `Cryptor` object the same way `CryptoPeerListener` already does, instead of trying to create a `Cryptor` instance every time `sign` and `encrypt` are called, the `Cryptor` object initialization doesn't seem to be trivial (ECC curves, etc.).

File: node/market.py [hunk 7]
```diff
@@ -658,26 +661,20 @@ def query_page(self, find_guid, callback=lambda msg: None):
...
-        )
-
+        peer = self.dht.routing_table.get_contact(msg['senderGUID'])
         def send_page_query():
             """Send a request for the local identity page"""
-            new_peer.start_handshake()
 
-            new_peer.send(proto_page(
+            self.log.debug('Sending page')
+            peer.send(proto_page(
```
Comment (gubatron) [inline node/market.py]: this inner `send_page_query()` function inside `on_query_page` seems unnecessary, should just remove and leave its code directly inside `on_query_page`

File: node/transport.py [hunk 8]
```diff
@@ -111,6 +118,17 @@ def __init__(self, ob_ctx, db_connection):
         if ob_ctx.enable_ip_checker and not ob_ctx.seed_mode and not ob_ctx.dev_mode:
             self.start_ip_address_checker()
 
+    def start_mediation(self, guid):
+        self.log.debug('Starting mediation %s', self.ob_ctx)
+        if self.ob_ctx.mediator:
+            for peer in self.dht.active_peers:
+                if peer.hostname == '205.186.156.31':
```
Comment (gubatron) [inline node/transport.py]: who is this IP?

File: node/openbazaar.py [hunk 9]
```diff
@@ -70,7 +71,8 @@ def create_argument_parser():
         ('--disable-stun-check',),
         ('--disable-upnp', '-j'),
         ('--enable-ip-checker',),
-        ('--seed-mode', '-S')
+        ('--seed-mode', '-S'),
+        ('--mediator', '-m')
```
Comment (gubatron) [inline node/openbazaar.py]: is this for UDP relaying? (proxy)

File: node/transport.py [hunk 10]
```diff
@@ -203,12 +266,125 @@ def _connect_to_bitmessage(self):
...
+                }))
+
+                peer2.send_raw(json.dumps({
+                    'type': 'punch',
+                    'guid': peer1.guid,
+                    'hostname': peer2.hostname,
+                    'port': peer2.port
+                }))
+            else:
+                ioloop.IOLoop.instance().call_later(1, send_punches)
```
Comment (gubatron) [inline node/transport.py]: is that `1` a second?

**Comment by Renelvon** [comment]
This PR touches so many files, that every other PR will inevitably conflict with this one. This cannot go on. It appears that @gubatron has already done his review and I trust he has done a good job, since networking is his expertise. I admit I have not done my part, as I had to shift focus away from the project and towards my graduation. To be completely honest, I don't feel like reviewing a 2.2K diff. However, from what I have gathered, this PR works, more or less. Considering the broken state of our networking, it seems that it would be best if this PR is merged ASAP. I propose we do so and fix code and style issues afterwards.

If anyone wants to self-assign this PR, please do.

**Comment by hoffmabc** [comment]
@Renelvon I plan to resolve as many issues as possible for this one, but it's a behemoth and will require much review going forward. There is no alternative at this point.

**Comment by gubatron** [comment]
@hoffmabc the next suggestion should be done sooner rather than later.

Try to think of adding a protocol version field somewhere on all of your current messages (unless you're planning to have this as a release that will be completely incompatible with future releases that define a more formal p2p messaging protocol)

This way in the future, you can refactor the code to be backwards compatible by parsing messages in different ways depending on the version number. I'd leave 2 bytes, for this, so you can have 255x255 possible versions of the protocol :)  0.0, 0.1, 0.255... 1.0, 1.1... and so on.

**Comment by hoffmabc** [comment]
We already have a 'v' field in the JSON that was supposed to be for this purpose. I need to go back and make sure it shows up.

**Comment by gubatron** [comment]
beautiful!

**Pull request closed without merging**